Шумилин Илья

Jevvry opened the pull request:
@LevShisterov

Jevvry committed:
Базовое решение

Jevvry committed:
Базовое решение

Jevvry committed:
Дополнение базового решения

Jevvry committed:
tests:
    - add exception handling tests

Jevvry committed:
feature:
    - add invalid font error handling
    - add cloud out of image size error handling

Jevvry committed:
tests:
    - add invalid font error handling test
    - add cloud out of image size error handling test

Jevvry committed:
remove unnecessary using

LevShisterov commented on TagsCloud/Extensions/RectangleExtension.cs:
```diff
@@ -0,0 +1,27 @@
+            return rectangle;
+        }
+
+        public static bool TryMoveRectangle(this Rectangle rectangle, Point delta, IEnumerable<Rectangle> shouldNotIntersect)
```
Похоже используется только в тестах.

LevShisterov commented on TagsCloud/Factory/IServiceFactory.cs:
```diff
@@ -0,0 +1,13 @@
+    {
+        Result<TService> Create();
+        IEnumerable<string> GetServiceNames();
+        IServiceFactory<TService> Register(string serviceName, Func<TService> wordConverter);
```
`wordConverter` - название не то.

LevShisterov commented on TagsCloud/Factory/ServiceFactory.cs:
```diff
@@ -0,0 +1,22 @@
+
+        public abstract Result<TService> Create();
+
+        public IEnumerable<string> GetServiceNames() => services.Select(pair => pair.Key);
```
`services.Keys`

LevShisterov commented on TagsCloud/ImageProcessing/Config/ImageConfig.cs:
```diff
@@ -0,0 +1,10 @@
+
+namespace TagsCloud.ImageProcessing.Config
+{
+    public class ImageConfig
```
Как думаешь, если я захочу иметь общий класс настроек, что мне делать? Если чуть точнее сформулировать: можно мне позволить не зависеть от этой конкретной реализации?

LevShisterov commented on TagsCloud/ImageProcessing/SaverImage/ImageSavers/BmpSaver.cs:
```diff
@@ -0,0 +1,11 @@
+    {
+        public override ImageFormat ImageFormat => ImageFormat.Bmp;
+
+        public override bool CanSave(string path) => path.EndsWith(".bmp");
```
А если имя будет в верхнем регистре?

LevShisterov commented on TagsCloud/ImageProcessing/SaverImage/Factory/ImageSaverFactory.cs:
```diff
@@ -0,0 +1,25 @@
+        {
+            var saver = services.FirstOrDefault(pair => pair.Value().CanSave(imageConfig.Path)).Value;
+
+            return Result.Of(saver, $"This image type {imageConfig.Path.Split('.').Last()} is not supported");
```
`System.IO.Path.GetExtension`

LevShisterov commented on TagsCloud/ImageProcessing/SaverImage/SaverBase.cs:
```diff
@@ -0,0 +1,23 @@
+            if (bitmap == null)
+                return Result.Fail<None>("Image was null");
+            var result = Result.OfAction(() => bitmap.Save(path, ImageFormat));
+            bitmap.Dispose();
```
Не очень хорошо что объект создается в одном классе, а диспозится в другом. Как это решить?

LevShisterov commented on TagsCloud/Layouter/TargetVector.cs:
```diff
@@ -0,0 +1,44 @@
+
+namespace TagsCloud.Layouter
+{
+    public class TargetVector
```
Похоже дельты сначала смещаются по 45 градусов линии, до ближайшей прямой и дальше идут по ней (при этом возвращая через раз `new Point(0, 0);`.
Действительно ли для этого класса было запланировано такое поведение?

LevShisterov commented on TagsCloud/TagsCloudProcessing/TagsGenerators/TagsGenerator.cs:
```diff
@@ -0,0 +1,60 @@
+                        .Take(count)
+                        .Select((wordInfo, index) =>
+                        {
+                            var font = new Font(wordsConfig.Font.FontFamily, count - index / 2);
```
Каждый `Font` - это `HANDLE` в Windows. Кэш? Dispose?

LevShisterov commented on TagsCloudTest/CircularCloudLayouterTests.cs:
```diff
@@ -0,0 +1,107 @@
+            var minSideSize = 10;
+
+            var sizes = new Size[length];
+            var random = new Random();
```
Каждый запуск будет разная последовательность, что может приводить к тесту, который то работает, то нет. Есть способ порешать это?

LevShisterov commented on TagsCloudTest/TagsCloudCreatorTests.cs:
```diff
@@ -0,0 +1,182 @@
+            var path = TestContext.CurrentContext.TestDirectory;
+
+            textPath = Path.Combine(path, "Resources", "text.txt");
+            imagePath = Path.Combine(path, "Resources", "test.png");
```
Этот комментарий к самому файлу "test.png". Имхо, его не должно быть в репозитории, а должен он быть в `.gitignore`.

Jevvry committed:
refactor:
    - add config Interfaces
    - ignore case in Readers, Savers
    - resolve IDisposable object

Jevvry committed:
tests:
    - code style

Jevvry committed:
refactor:
    - remove TryMoveRectangle in RectangleExtensions
    - remove unnecessary using

Jevvry committed:
refactor:
    - remove unnecessary designer file

Jevvry committed:
feature:
    - add return to the menu in ConfigWindow

Jevvry committed:
preparing for migration to  .net core 3.1

Jevvry committed:
add test.png to .gitignore

Jevvry committed:
migration to .net core 3.1
